src/core/lib/viewport-transition-interpolator.js

```diff
@@ -0,0 +1,185 @@
+import {projectFlat, unprojectFlat} from 'viewport-mercator-project';
+import {Vector2, equals} from 'math.gl';
+import assert from 'assert';
+
+import {isValid, isAngularProp} from './viewport-transition-utils';
+
+const EPSILON = 0.01;
+
+const VIEWPORT_TRANSITION_PROPS =
+  ['longitude', 'latitude', 'zoom', 'bearing', 'pitch'];
+
+/** Util functions */
+function lerp(start, end, step) {
+  if (Array.isArray(start)) {
+    return start.map((element, index) => {
+      return lerp(element, end[index], step);
+    });
+  }
+  return step * end + (1 - step) * start;
+}
+function zoomToScale(zoom) {
+  return Math.pow(2, zoom);
+}
+
+function scaleToZoom(scale) {
+  return Math.log2(scale);
+}
+
+export class ViewportTransitionInterpolator {
+
+  /**
+   * Checks if two sets of props need transition in between
+   * @param currentProps {object} - a list of viewport props
+   * @param nextProps {object} - a list of viewport props
+   * @returns {bool} - true if two props are equivalent
+   */
+  arePropsEqual(currentProps, nextProps) {
+    for (const key of VIEWPORT_TRANSITION_PROPS) {
+      if (!equals(currentProps[key], nextProps[key])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns viewport props in transition
+   * @param startProps {object} - a list of starting viewport props
+   * @param endProps {object} - a list of target viewport props
+   * @param t {number} - a time factor between [0, 1]
+   * @returns {object} - a list of interpolated viewport props
+   */
+  interpolateProps(startProps, endProps, t) {
+    // Dummy return value -- this should be overriden by child classes
+    return endProps;
+  }
+
+}
+
+/**
+ * Performs linear interpolation of two viewports.
+*/
+export class ViewportLinearInterpolator extends ViewportTransitionInterpolator {
```

Comment:
How close is this to being a generic `LinearInterpolator`? Looks like this could be a subclass that just informs the superclass which props are angles?